### ISSUE TITLE
The current state of `Dispatcher` contentions

### ISSUE THREAD

**Issue by danicheg** [typelevel/cats-effect#2933]
This issue is a follow-up from the discussion https://github.com/http4s/http4s/pull/6212.
`Dispatcher`'s scaladoc [contains](https://github.com/typelevel/cats-effect/blob/27307cfe6cd9ea3cb230a8f9be5e16f624a914cd/std/shared/src/main/scala/cats/effect/std/Dispatcher.scala#L40-L42) advice to create instances of it as often as possible to prevent possible contentions (very roughly speaking). Is it still so? Or maybe things have changed?

**Comment by djspiewak** [comment]
This desperately needs better documentation :-(

The tldr is that a single global dispatcher is completely fine if you don't care about sequentialization, especially since #2876. However, one dispatcher per request is also fine and arguably *better* if you either care about sequentiality, or if you know a priori that you only have a single producing thread, but only once #2854 lands.

**Comment by wjoel** [comment]
I think the "cancel all side effects when the `Dispatcher` instance goes away, even if the user asked us to just run and forget about them" aspect of `Dispatcher` is part of it. It makes sense for `Supervisor`, but it's surprising behavior for `Dispatcher`, and I suspect it's the main reason for https://github.com/typelevel/cats-effect/issues/2727

Consider this example.
```scala
import cats.effect.std.Dispatcher
import cats.effect.{Deferred, IO, IOApp}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration.DurationInt

object Main extends IOApp.Simple {
  private def getJavaAsyncResult(resultSuccess: Deferred[IO, String]): IO[Unit] =
    // We're at an impure boundary, so we should use a Dispatcher, and we're encouraged
    // to create many of them to avoid contention, so this is what we should do, right?
    Dispatcher[IO].use { dispatcher =>
      // Pretend that we're doing a database call, sending an email using the sendgrid Java SDK, or whatever,
      // not just sleeping.
      IO(Future(Thread.sleep(500))
        .recover(_ => dispatcher.unsafeRunAndForget(resultSuccess.complete("failure")))
        .onComplete(_ => dispatcher.unsafeRunAndForget(resultSuccess.complete("success")))
      )
    }.void

  override def run: IO[Unit] =
    for {
      _ <- IO.println("Started")
      deferredResult <- Deferred[IO, String]
      _ <- getJavaAsyncResult(deferredResult)
      _ <- IO.println("Waiting for the future")
      result <- deferredResult.get
    } yield println(s"Finished with $result")
}
```

To me (by now, after seeing enough "dispatcher already shutdown" errors) and you it may not be surprising that it fails with "dispatcher already shutdown", but to anyone new to cats-effect, I think it is, and in real programs it may not happen so immediately and can be very difficult to troubleshoot.

The `.complete` actions could, of course, be replaced with log4cats logging, epimetheus metric updates, or whatever - the deferred itself isn't important, the important thing is that if you have one dispatcher per request you better be sure that you're ok with `.unsafeRunAndForget` actions getting canceled when the request gets canceled.

Changing it to this, it all works out ok, but it goes against the advice to create many dispatchers, and has nothing to do with performance as in the http4s issue.
```scala
object Main extends IOApp.Simple {
  private def getJavaAsyncResult(dispatcher: Dispatcher[IO], resultSuccess: Deferred[IO, String]): IO[Unit] =
    IO(Future(Thread.sleep(500))
      .recover(_ => dispatcher.unsafeRunAndForget(resultSuccess.complete("failure")))
      .onComplete(_ => dispatcher.unsafeRunAndForget(resultSuccess.complete("success")))
    )

  override def run: IO[Unit] = {
    Dispatcher[IO].use { dispatcher =>
      for {
        _ <- IO.println("Started")
        deferredResult <- Deferred[IO, String]
        _ <- getJavaAsyncResult(dispatcher, deferredResult)
        _ <- IO.println("Waiting for the future")
        result <- deferredResult.get
      } yield println(s"Finished with $result")
    }
  }
}
```

Part of this is just a documentation issue (seems we all agree it should be changed), but to to fix this properly we may need to consider changing the semantics of `Dispatcher`, at least for `unsafeRunAndForget`. It just doesn't work as expected with async Java APIs, in my opinion.

If I was using a `Supervisor` directly, the behavior of the first example above isn't (shouldn't be) surprising as it "spawns fibers whose lifecycles are bound to that of the supervisor" according to the documentation, but if I'm just using a `Dispatcher` because it's what I should use "for evaluating effects across an impure boundary" and it has an `unsafeRunAndForget` method, I think it is. Which is probably why we have https://github.com/typelevel/cats-effect/issues/1881, as well.

**Comment by djspiewak** [comment]
I want to add some more configuration to `Dispatcher` (#1881). Honestly, #2854 decently opens the door to doing this, since it adjusts the construction API. Backpressuring rather than canceling on completion is definitely a thing we should allow for, and maybe even default.
